zephyr: switch over to a simple priority-based LL scheduler

lyakh opened the pull request:
This replaces the original LL scheduler with a simplified version.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+ * checks, but when we switch over to drop the spin-lock when running tasks,
+ * we'll have to come up with a robust implementation.
+ */
+static int zephyr_ll_task_reschedule(void *data, struct task *task, uint64_t start)
```
ack, lets just say -ENOTSUP here as DMA trace will be using a thread with Zephyr.

lgirdwood commented on zephyr/CMakeLists.txt:
```diff
@@ -393,7 +393,7 @@ zephyr_library_sources(
 	${SOF_SRC_PATH}/schedule/dma_single_chan_domain.c
 	${SOF_SRC_PATH}/schedule/dma_multi_chan_domain.c
-	${SOF_SRC_PATH}/schedule/ll_schedule.c
+	${SOF_SRC_PATH}/schedule/zephyr_ll.c
```
Lets make this a Kconfig option - both will export the same scheduler type.

lyakh replied:
@lgirdwood you want both to be available with Zephyr - selectable via Kconfig?

lgirdwood replied:
ah, got you - this in in the zephyr makefile. Please ignore.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+			continue;
+		}
+
+		if (!domain_is_pending(sch->ll_domain, task, &sched_comp))
```
what does this check ?

lyakh replied:
that the task's own period has expired

lyakh replied:
this should be removed together with `.start` and `.next_tick` in a follow-up, I'll add a comment

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+			continue;
+
+		task->state = SOF_TASK_STATE_RUNNING;
+		state = task_run(task);
```
We should really release the spinlock before we run() and then acquire it again after the run_task() returns.

lyakh replied:
we have to research well what contexts can race for which resources here. IIUC and scheduling can be triggered from interrupt handlers, then the list of tasks has to be protected. In that case we cannot just drop the lock in the middle of this loop. One solution would be what I proposed - first extract all runnable tasks from the global list into a local one, while holding the lock. Then run the tasks from the local list, with no locking this time. The disadvantage would be, that tasks, returning `RESCHEDULE` have to be re-added to the global list, but I don't have a better solution so far.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+	task->start = domain_time(sch) + delay;
+	if (sch->ll_domain->next_tick != UINT64_MAX &&
+	    sch->ll_domain->next_tick > task->start)
+		task->start = sch->ll_domain->next_tick;
```
I dont think we need to have the concept of task start time here as the task will only start at the next LL tick iff it's in the task list.

lyakh replied:
let's look at this once it's first working reliably as is, I'll want to understand this better.

lgirdwood replied:
I assume this is removed in the subsequent PR ?

lyakh replied:
it will be, yes, adding comments for that

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+	/*
+	 * FIXME: similar to zephyr_ll_task_reschedule() - what do we do when
+	 * the scheduling loop isn't holding the lock any more?
+	 */
```
Is this LL tick running at this point or not ? (as this impacts the complexity)

lyakh replied:
it should be, yes

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+	 * check there will be protected by the spin-lock. Then we can take the
+	 * spin-lock here too. Doing this now can potentially deadlock.
+	 */
+	task->state = SOF_TASK_STATE_CANCEL;
```
I think this is probably fine.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+
+	spin_unlock(&sch->lock);
+
+	ret = domain_register(sch->ll_domain, task, &zephyr_ll_run, sch);
```
What does this do agian, I guess we also need the unregister too ?

lyakh replied:
oops, indeed we do, thanks for spotting this! In Zephyr case it starts and stops the thread for the first and the last task respectively, and starts and stops the timer for the first and the last task globally over all cores

lgirdwood reviewed (commented):
Good stuff, it's a great simplification. It's also obvious where we can improve the external and internal APIs here.
One thing we do need to look at next week is when we add a LL task we probably also need to add LL synchronized logic that does the triggering too.

kv2019i reviewed (commented):
Much easier to follow. Probably the hardest parts relate to places where the "domain" abstraction is used to call into zephyr_domain.c.

lyakh commented on src/include/sof/schedule/ll_schedule.h:
```diff
@@ -34,11 +34,24 @@ struct ll_task_pdata {
+#define scheduler_init_ll zephyr_ll_scheduler_init
+#define schedule_task_init_ll zephyr_ll_task_init
+
+#endif
```
I first implemented both with keeping the "old" names, but then I thought that having proper namespace consistency in the .c file is better... But we can discuss this.

lyakh commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,296 @@
+};
+
+//FIXME
+#define domain_time(sch) platform_timer_get_atomic(timer_get())
```
well, they have 1 *zephyr* tick granularity, i.e. 50 or 48kHz

lyakh replied:
have to change it in zephyr_domain.c too then

lyakh commented on src/schedule/zephyr_domain.c:
```diff
@@ -120,6 +126,8 @@ static int zephyr_domain_register(struct ll_schedule_domain *domain,
 
 		k_timer_start(&zephyr_domain->timer, start, K_USEC(LL_TIMER_PERIOD_US));
+		domain->next_tick = platform_timer_get_atomic(zephyr_domain->ll_timer) +
+			k_ticks_to_cyc_ceil64(k_timer_remaining_ticks(&zephyr_domain->timer));
```
let's keep `.next_tick` and `.start` in this PR, I'll add a comment that they shouldn't be needed for a fully native Zephyr LL scheduling stack, and let's try to remove them in a follow up PR.

lyakh commented:
The latest update addresses comments and also fixes the Zephyr multicore case (where supported), which got broken with the new scheduler because it didn't account for a self-terminating thread

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+		task->priority, task->flags);
+
+	irq_local_disable(flags);
+	if (!sch->n_tasks)
```
can we have some {} or the comment above the if() otherwise it looks like the indentation is wrong.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+		 * While the lock is not held, the task can be cancelled, we
+		 * deal with it after the task completion.
+		 */
+		if (must_run) {
```
why not just check task->state here ?

lyakh replied:
having removed `PENDING` it can be done, yes

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+		 * the task is cancelled, which is an intrinsic race, that
+		 * cannot be prevented as long as we want to run tasks unlocked.
+		 */
+		task->state = SOF_TASK_STATE_PENDING;
```
if this is optional then lets remove this state change here.

lyakh replied:
ok

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,399 @@
+		task = container_of(list, struct task, list);
+
+		spin_lock_irq(&sch->lock, flags);
+		if (task->state == SOF_TASK_STATE_PENDING) {
```
at this point we are **committed** to run everything in the list, so this can be simplified.

lyakh replied:
we know, that tasks can take a relatively long time to execute. If you have multiple tasks on the (temporary) scheduler list and you start executing the first of them, then the next one... There's a rather large window when tasks, further down on the list, can be cancelled. Why not use the chance and avoid running them, saving some execution cycles?

lgirdwood replied:
This is fine, as it simplifies the execution path and we will always be faster than the IPC time window

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+
+/*
+ * This is synchronous - after this returns the object can be destroyed!
+ * Assertion: under Zephyr this is always called from a thread context!
```
Lets make this runtime check and shout if we are in IRQ context.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+	bool must_wait;
+
+	if (k_is_in_isr())
+		k_panic();
```
Lets not panic, just complain and return. Panic stops the trace.

lyakh replied:
ok

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+		 * zephyr_ll_task_done() before aborting the thread, so just
+		 * disabling local IRQs provides sufficient protection here.
+		 */
+		spinlock_init(&sch->lock);
```
Why is the spinlock not automatically initialized at core boot as part of scheduler init ?

lyakh replied:
because it has to be initialised every time a core submits the first task for scheduling

lgirdwood replied:
This is broken and need fixed.

lyakh replied:
the reason is, that when the last task on a core completes, the LL-scheduling thread on that core terminates *while holding the spin-lock*. And the spin-lock is a part of per-core scheduler data, so, next time we start the scheduling thread on that core we re-initialise the spin-lock. So far I don't see a sufficiently clean way to eliminate this... Is it really that bad?

lgirdwood replied:
Yes, it's bad - lets at least put it int a static inline away from the main flow (for easier reading)

lyakh replied:
you're proposing a
```
static zephyr_ll_init_scheduler_for_first_task(struct zephyr_ll *sch)
{
	spinlock_init(&sch->lock);
}
```
? Yeah, maybe that would self-document it a bit and add a scope to it too.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+		/* keep original start */
+		if (task_iter == task) {
+			spin_unlock_irq(&sch->lock, flags);
+			return 0;
```
Should we not complain here and  return an error ? Under what circumstances would be schedule a task twice ?

lyakh replied:
Don't know, this is taken from the original ll-scheduler. We can try to remove it and see what breaks, but I'd rather do it later in a separate PR. Let me add a comment here.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,399 @@
+
+		pdata = task->priv_data;
+
+		if (pdata->terminator) {
```
ok, terminator is confusing - this is really a zephyr thread of the waiting thread ?

lyakh replied:
this is the thread, that has called `task_free()`, yes. Freeer is ugly, liberator is pathetic :-) `freeing_thread` is verbose and clear, but long?

lyakh commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,435 @@
+	domain_unregister(sch->ll_domain, task, --sch->n_tasks);
+}
+
+/* The caller must hold the lock and possibly disable interrupts */
```
this is a static internal function, maybe it's better to add such checks to callers, if any of them is uncertain

lgirdwood commented on src/schedule/zephyr_domain.c:
```diff
@@ -70,11 +71,16 @@ static void zephyr_domain_thread_fn(void *p1, void *p2, void *p3)
 	if (!zephyr_domain)
 		return;
 
+	/* This loop should only run once */
```
I guess this loop can be removed in the next PR ?

lgirdwood commented on src/include/sof/schedule/ll_schedule_domain.h:
```diff
@@ -142,19 +142,29 @@ static inline void domain_unregister(struct ll_schedule_domain *domain,
+		domain->registered[core] = false;
+
+	/*
+	 * In some cases .domain_unregister() might not return, terminating the
```
Can we state these cases in the comment.

lgirdwood commented on src/schedule/zephyr_domain.c:
```diff
@@ -111,6 +114,9 @@ static int zephyr_domain_register(struct ll_schedule_domain *domain,
 	dt->handler = handler;
 	dt->arg = arg;
 
+	/* 10 is rather random, we better not accumulate 10 missed timer interrupts */
```
Lets make this a macro too.

lgirdwood commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,455 @@
+	struct ll_schedule_domain *ll_domain;	/* scheduling domain */
+};
+
+/* per-task scheduler data: so far actually unused */
```
Should this be in the next PR ?

lyakh replied:
Actually this data is already used, I'll update the comment

lgirdwood reviewed (commented):
Looks like some of these can do with a squash

kv2019i commented on src/schedule/zephyr_domain.c:
```diff
@@ -70,11 +71,16 @@ static void zephyr_domain_thread_fn(void *p1, void *p2, void *p3)
 
+	/* This loop should only run once */
+	while (zephyr_domain->ll_domain->next_tick < now)
+		zephyr_domain->ll_domain->next_tick += LL_TIMER_PERIOD_TICKS;
```
Not sure if we need to finetune it now (if next_tick is removed later), but this looks a bit dangerous. Can next_tick be zero and this end up spinning for a longer time?

kv2019i commented on src/schedule/zephyr_ll.c:
```diff
@@ -0,0 +1,455 @@
+ * RUNNING:	the scheduler is running, the task is moved to a temporary list
+ *		and then executed
+ * CANCEL:	anywhere, where the lock is not held, tasks can be cancelled
+ *		asynchronously from an arbitrary context
```
The above text for CANCEL is not really describing the state, but rather the transition to cancel.

kv2019i reviewed (commented):
Another look, the main functions look good. No major new issues found (a few minor ones inline).

lgirdwood reviewed (approved):
Ok, as this provides a stable base point we can merge and address the improvents and simplifications in subsequent PRs.
@keqiaozhang will this now let us test in CI ?

marc-hb commented:
I have to revert this PR locally because when I re-enable the `trace_work()` thread in Zephyr (with PR #4452) this new scheduler ignores `DMA_TRACE_PERIOD` and seems to re-run it as fast as it can. I bisected this to commit `zephyr: ll-schedule: switch over to a simplified implementation`. The same commit adds the following error message which could  be relevant?
```
zephyr_ll_scheduler_init(): unsupported domain 2
```
This is on APL Up Squared.

lgirdwood commented:
@lyakh any inputs ?

lyakh commented:
> I have to revert this PR locally because when I re-enable the `trace_work()` thread in Zephyr (with PR #4452) this new scheduler ignores `DMA_TRACE_PERIOD` and seems to re-run it as fast as it can. I bisected this to commit `zephyr: ll-schedule: switch over to a simplified implementation`. The same commit adds the following error message which could be relevant?
> 
> ```
> zephyr_ll_scheduler_init(): unsupported domain 2
> ```
> 
> This is on APL Up Squared.

@marc-hb @lgirdwood Right, the current Zephyr LL scheduler version runs all tasks with the same period (1ms by default). We can change that if needed. @lgirdwood proposed to use a counter for that. But I'm also wondering - does the DMA trace task have to be LL or should it rather be EDF?

marc-hb commented:
>  does the DMA trace task have to be LL or should it rather be EDF?

I don't know why it has to be LL.

lgirdwood commented:
Trace should be preemptable, is should be DP (aka EDF).
